redis/src/cluster_async/mod.rs

```diff
@@ -970,15 +970,29 @@ where
         // When we no longer need to support Rust versions < 1.67, remove fast_math and transition to the ilog2 function.
         let num_of_nodes_to_query =
             std::cmp::max(fast_math::log2_raw(num_of_nodes as f32) as usize, 1);
-        let (_, found_topology_hash) = calculate_topology_from_random_nodes(
+        let (res, failed_connections) = calculate_topology_from_random_nodes(
             &inner,
             num_of_nodes_to_query,
             &read_guard,
             DEFAULT_NUMBER_OF_REFRESH_SLOTS_RETRIES,
         )
-        .await?;
-        let change_found = read_guard.get_current_topology_hash() != found_topology_hash;
-        Ok(change_found)
+        .await;
+
+        if let Ok((_, found_topology_hash)) = res {
+            if read_guard.get_current_topology_hash() != found_topology_hash {
+                return Ok(true);
+            }
+        }
+        drop(read_guard);
+
+        Self::refresh_connections(
```

Comment:
You haven't included in this PR saving connections that were failed to be refreshed and retrying to refresh them on the next periodic task. It means that we might end up only with user connections, as all management connections will be removed when we'll fail to refresh them, and they won't be updated until a real topology change will happen. I think that we should do the best not to use the user connections for periodic checks, so retrying to refresh them is important, as I've done in the PR here:
https://github.com/amazon-contributing/redis-rs/pull/56/files#diff-5ddfd8b2718c21504d43c50b0490bc22de4189208d97f5604a2297124e529577R871

Reply:
The solution suggested in your PR isn't complete either - it only registers errors on the periodic check, not on triggered topology updates.

If you have no comments about this change, let's move forward with this, and afterwards try to find a complete solution *which we're able to test*. Let's try to keep PR scopes limited, and not repeat the error of building large, complex logic without tests.

Reply:
BTW, test suggestions for this PR will be appreciated. I'm stumped 😅

Reply:
As you wrote that this is the final item, I mentioned we still haven't covered failed reconnections